### DISCUSSION TITLE
SSR and Server-Side Helpers for App Router

### DISCUSSION THREAD

**Discussion by tomitrescak** [trpc/trpc#5674] [Q&A]
I am using TRPC with app router. I like the way TRCP handles data updates on the clients, and I wouldn't say I like mutations in server-side components. Thus, I consciously decided to render all non-changeable data in the server component but fetch all the modifiable data using the client component and topic/react approach.

But, I really dislike the "Loading ..." interface and I want to have SSR. I am using the T3-Stack. The problem is that I could not figure out how to do SSR, as all the documentation points to the old app router.

There are https://trpc.io/docs/client/nextjs/server-side-helpers, but the docs only mention pages router, and no matter how I tried to bend it, the TRPC cache was not hydrated. The t3-stack from the app router also seems not to use the @trcp/next component anymore.

**Comment by KenSoftDev** [comment]
https://github.com/trpc/trpc/discussions/5108#discussioncomment-9299904

**Comment by rsslldnphy** [comment]
I have the same question, and not sure I understand how the linked comment relates to it?

**Comment by marcelo-mason** [comment]
Set up your SSR Helper:

**ssr.ts**
```tsx
"use server"

import { createServerSideHelpers } from "@trpc/react-query/server"
import { AppRouter, appRouter } from "@/server/trpc/root"
import { createInnerTRPCContext } from "@/server/trpc/trpc"

/**
 * SSR helpers (for prefetching data on the server)
 * https://trpc.io/docs/client/nextjs/server-side-helpers
 */
export const createSSRHelper = async () => {
  const session = await getSession()
  return createServerSideHelpers<AppRouter>({
    router: appRouter,
    ctx: createInnerTRPCContext({ session }),
    transformer: SuperJSON,
  })
}
```

Set up your Hydration server component.  Prefetch the queries you want to pass to the frontend:

**Hydration.ts**
```tsx
"use server"

import { dehydrate, HydrationBoundary } from "@tanstack/react-query"
import { createSSRHelper } from "@/server/trpc/ssr"

export async function Hydration({ children }: ChildrenProp) {
  const ssr = await createSSRHelper()

  // prefetches
  await ssr.articles.getPopular.prefetch()
  await ssr.topics.getFirstPage.prefetch()

  const dehydratedState = dehydrate(ssr.queryClient)
  return (
    <HydrationBoundary state={dehydratedState}>{children}</HydrationBoundary>
  )
}
```

Add it to your page's layout :

**layout.ts**
```tsx
export default async function RootLayout({ children }: ChildrenProp) {

  return (
    <html>
      <body>
        <main>
          <Hydration>{children}</Hydration>
        </main>
      </body>
    </html>
  )
}
```
